simple-ipc--unix-sockets.c

```diff
@@ -170,4 +170,898 @@ int ipc_client_send_command(const char *path,
+ * rather just large enough for our "accept-thread" to wake up and
+ * queue incoming connections onto the FIFO without the kernel
+ * dropping any.
+ *
+ * This value chosen at random.
+ */
+#define LISTEN_BACKLOG (50)
+
+/*
+ * Create a unix domain socket at the given path to listen for
+ * client connections.  The resulting socket will then appear
+ * in the filesystem as an inode with S_IFSOCK.  The inode is
+ * itself created as part of the `bind(2)` operation.
+ *
+ * Unix domain sockets have a fundamental design flaw because the
+ * inode persists until the pathname is deleted; closing the listening
+ * socket only closes the socket handle/descriptor, it does not delete
+ * the inode.
+ *
+ * Well-behaving service daemons are expected to also delete the inode
+ * before shutdown.  If a service crashes (or forgets) it can leave
+ * the (now stale) inode in the filesystem.  This behaves like a stale
+ * ".lock" file and may prevent future service instances from starting
+ * up correctly.
+ *
+ * When future service instances try to create the listener socket,
+ * `bind(2)` will fail with EADDRINUSE -- because the inode already
+ * exists.  However, the new instance cannot tell if it is a stale
+ * inode *or* another service instance is already running.
+ *
+ * One possible solution is to blindly unlink the inode before
+ * attempting to create the new socket and inode.  Then `bind(2)`
+ * should always succeed.  However, if there is an existing service
+ * instance, it would be orphaned -- listening on a socket/inode that
+ * is no longer visible to future clients.
+ *
+ * An alternative solution is to gently create the socket.  If the
+ * inode already exists, try to determine whether it is stale or
+ * actually in use by an existing service process by attempting to
+ * connect to the socket/inode.  (This assumes that the existing
+ * process isn't wedged.)  Then if stale, recreate the socket with
+ * force.
+ */ 
+static int create_listener_socket(const char *path)
+{
+	int fd_listen;
+	int fd_client;
+	struct unix_stream_listen_opts opts = {
+		.listen_backlog_size = LISTEN_BACKLOG,
+		.force_unlink_before_bind = 0,
+		.disallow_chdir = 1
+	};
+
+	trace2_data_string("ipc-server", NULL, "try-listen-gently", path);
+
+	fd_listen = unix_stream_listen_gently(path, &opts);
+	if (fd_listen >= 0)
+		return fd_listen;
+
+	if (errno != EADDRINUSE)
+		return error_errno(_("could not create socket '%s'"),
+				   path);
+
+	trace2_data_string("ipc-server", NULL, "try-detect-server", path);
+
+	fd_client = unix_stream_connect(path);
+	if (fd_client >= 0) {
+		/*
+		 * An existing service process is alive and accepted our
+		 * connection.
+		 */
+		close(fd_client);
+
+		errno = EADDRINUSE;
+		return error_errno(_("socket already in use '%s'"),
+				   path);
+	}
+
+	trace2_data_string("ipc-server", NULL, "try-listen-force", path);
+
+	opts.force_unlink_before_bind = 1;
+	fd_listen = unix_stream_listen_gently(path, &opts);
```

Comment:
There is a race here, right? Between `unix_stream_connect()` and `unix_stream_listen_gently()`, another service process could have _just_ started listening. Maybe clarify that in an inline comment? (To me, the comment above the function does not really clarify this point.)

Reply:
i've completely reworded the comments above and in-line in that function.  let me know if that helps.